torch/nn/modules/conv.py

```diff
@@ -636,6 +636,102 @@ def forward(self, input, output_size=None):
             output_padding, self.groups, self.dilation)
 
 
+class Conv2dBackward(_ConvTransposeMixin, _ConvNd):
```

Comment:
I don't think that adding this module is a good idea. First, as noticed in this docstring, it's pretty much just conv transpose. Secondly, once we have `Conv2dBackward`, why shouldn't we also add `Conv3dBackward`, `LinearBackward`, `ReLUBackward`, etc...

What's the use case for this? Maybe we should just have an interface for accessing the gradients we implement in autograd, but as a functional interface that doesn't need the first stage to happen and capture the relevant elements?

Reply:
@apaszke The issue #5024 discusses some use cases where they require alternate forms of backprop through convolution. This API is functionally similar to what TensorFlow's https://www.tensorflow.org/api_docs/python/tf/nn/conv2d_backprop_input provides. Tensorflow also uses the same underlying API for both conv2d_backprop_input and [conv2d_transpose](https://github.com/tensorflow/tensorflow/blob/54ab9d690a63ee11813f091d9dd9c2cd3c4cce1d/tensorflow/python/ops/nn_ops.py#L1183). 

Do you suggest we should make a generic interface that can call into the backward pass for any operator function?

Reply:
So, the real utility of exposing backwards convolution is not the backwards on input (which is pretty simple) but the backwards on weight, which is also itself a convolution but with dilation and stride swapped. Yes, the user can do the math themselves, but why not just make their life easier.

As for the slippery slope argument, I think the answer is the same as all the other ops; we add them if people ask for them, and don't if they don't. A functional interface to autograd sounds like an awfully convoluted way to do this. When we have the symbolic differentiator things are probably less bad.

@vedanuj We're probably going to need all of the other variants, but maybe we should first convince @apaszke this is a good idea :)

Reply:
I agree that having *some* way of supporting people who need this would be nice. I'm just saying that making the nn API 2x larger doesn't seem like a great solution, especially considering that it doesn't solve the original problem mentioned in that forum thread:

> I want to use a coustom forward function with the standard convolutional backward function.
> Maybe I should use the standard function when I need a backpropagation. Thank you.

Implementing a backward of an autograd function using a module seems like a bad idea to me. Adding it to functional would be better, but I still don't like the fact that it's another function we have to maintain manually, when we're auto-generating them for internal purposes.

Symbolic differentiation is never meant to be exposed to users, and is designed for internal use in the JIT only. We don't even plan to expand its scope to cover all of autograd in the core library.

Reply:
> Implementing a backward of an autograd function using a module seems like a bad idea to me. Adding it to functional would be better, but 

Having just the functional seems fine to me.

> I still don't like the fact that it's another function we have to maintain manually, when we're auto-generating them for internal purposes.

But we're not auto-generating them! We have explicitly implemented kernels for each of the backwards. We're just giving aliases to them to help users out.

Reply:
Fine, let's do the functional version only, like TF does it.